File: src/models/series.model.ts
export interface SeriesFetchModel {
    series_id: number;
    guid: string;
    title: string;
    description: string;
    preview_image: string | null;
}

export interface SeriesDbModel {
    series_id: number;
    guid: string;
    title: string;
    description: string;
    preview_image: string | null;
}

export interface SeriesModel extends SeriesDbModel {
    clone(): SeriesModel;
}

export function SeriesModel(guid: string, title: string, description: string, preview_image: string | null): SeriesModel;
export function SeriesModel(series_id: number, guid: string, title: string, description: string, preview_image: string | null): SeriesModel;
export function SeriesModel(...args: unknown[]): SeriesModel {
    if (args.length == 4) {
        args.unshift(0);
    }

    return _SeriesModel(...args as [number, string, string, string, string | null]);
}

function _SeriesModel(series_id: number, guid: string, title: string, description: string, preview_image: string | null): SeriesModel {
    const obj: SeriesModel = {
        series_id: series_id,
        guid: guid,
        title: title,
        description: description,
        preview_image: preview_image,
        clone: clone,
    };

    obj.clone = obj.clone.bind(obj);
    return obj;
}

function clone(this: SeriesModel): SeriesModel {
    const obj: SeriesModel = {
        series_id: this.series_id,
        guid: this.guid,
        title: this.title,
        description: this.description,
        preview_image: this.preview_image,
        clone: clone,
    };

    obj.clone = obj.clone.bind(obj);
    return obj;
}

File: server/AniStream/DTO/SeriesModel.cs
namespace AniStream.API.DTO;

public sealed class SeriesModel
{
    public required int SeriesId { get; set; }

    public required string Guid { get; set; }

    public required string Title { get; set; }

    public required string Description { get; set; }

    public required string? PreviewImage { get; set; }
}

public sealed class SeriesFilterModel
{
    public required int Offset { get; set; }

    public required int Limit { get; set; }

    public required string? SearchText { get; set; }

    public required int[]? GenreIds { get; set; }
}

public sealed class SeriesCreateModel
{
    public required string Guid { get; set; }

    public required string Title { get; set; }

    public required string Description { get; set; }

    public required string? PreviewImage { get; set; }
}

public sealed class SeriesSyncModel
{
    public required bool RequiresSync { get; set; }
    public required bool IsSyncing { get; set; }
}

internal static class SeriesModelHelper
{
    public static SeriesModel ToDTO(this Models.SeriesModel model)
    {
        return new SeriesModel
        {
            SeriesId = model.SeriesId,
            Guid = model.Guid,
            Title = model.Title,
            Description = model.Description,
            PreviewImage = model.PreviewImage
        };
    }
}